staging/src/k8s.io/api/resource/v1beta2/types.go

```diff
@@ -0,0 +1,1510 @@
+	//
+	// The maximum number of attributes and capacities combined is 32.
+	//
+	// +optional
+	Attributes map[QualifiedName]DeviceAttribute `json:"attributes,omitempty" protobuf:"bytes,2,rep,name=attributes"`
+
+	// Capacity defines the set of capacities for this device.
+	// The name of each capacity must be unique in that set.
+	//
+	// The maximum number of attributes and capacities combined is 32.
+	//
+	// +optional
+	Capacity map[QualifiedName]DeviceCapacity `json:"capacity,omitempty" protobuf:"bytes,3,rep,name=capacity"`
+
+	// ConsumesCounter defines a list of references to sharedCounters
+	// and the set of counters that the device will
+	// consume from those counter sets.
+	//
+	// There can only be a single entry per counterSet.
+	//
+	// The maximum number of device counter consumption entries
+	// is 32. This is the same as the maximum number of shared counters
+	// allowed in a ResourceSlice.
+	//
+	// +optional
+	// +listType=atomic
+	// +featureGate=DRAPartitionableDevices
+	ConsumesCounter []DeviceCounterConsumption `json:"consumesCounter,omitempty" protobuf:"bytes,4,rep,name=consumesCounter"`
+
+	// NodeName identifies the node where the device is available.
+	//
+	// Must only be set if Spec.PerDeviceNodeSelection is set.
+	// At most one of NodeName, NodeSelector and AllNodes can be set.
+	//
+	// +optional
+	// +oneOf=DeviceNodeSelection
+	// +featureGate=DRAPartitionableDevices
+	NodeName *string `json:"nodeName,omitempty" protobuf:"bytes,5,opt,name=nodeName"`
+
+	// NodeSelector defines the nodes where the device is available.
+	//
+	// Must use exactly one term.
+	//
+	// Must only be set if Spec.PerDeviceNodeSelection is set.
+	// At most one of NodeName, NodeSelector and AllNodes can be set.
+	//
+	// +optional
+	// +oneOf=DeviceNodeSelection
+	NodeSelector *v1.NodeSelector `json:"nodeSelector,omitempty" protobuf:"bytes,6,opt,name=nodeSelector"`
+
+	// AllNodes indicates that all nodes have access to the device.
+	//
+	// Must only be set if Spec.PerDeviceNodeSelection is set.
+	// At most one of NodeName, NodeSelector and AllNodes can be set.
+	//
+	// +optional
+	// +oneOf=DeviceNodeSelection
+	// +featureGate=DRAPartitionableDevices
+	AllNodes *bool `json:"allNodes,omitempty" protobuf:"bytes,7,opt,name=allNodes"`
+
+	// If specified, these are the driver-defined taints.
+	//
+	// The maximum number of taints is 8.
+	//
+	// This is an alpha field and requires enabling the DRADeviceTaints
+	// feature gate.
+	//
+	// +optional
+	// +listType=atomic
+	// +featureGate=DRADeviceTaints
+	Taints []DeviceTaint `json:"taints,omitempty" protobuf:"bytes,8,rep,name=taints"`
+}
+
+// DeviceCounterConsumption defines a set of counters that
+// a device will consume from a CounterSet.
+type DeviceCounterConsumption struct {
+	// SharedCounter defines the shared counter from which the
+	// counters defined will be consumed.
+	//
+	// +required
+	SharedCounter string `json:"sharedCounter" protobuf:"bytes,1,opt,name=sharedCounter"`
```

Comment:
This is the name of a CounterSet, not of a "shared counter".

A shared counter is a `Counter` instance, which is identified by the name of its set and the counter name.

```suggestion
	// CounterSet defines the set from which the
	// counters defined will be consumed.
	//
	// +required
	CounterSet string `json:"counterSet" protobuf:"bytes,1,opt,name=counterSet"`
```

cc @cici37 @johnbelamaric

Reply:
I stumbled over this when updating the validation code.

Populating `SharedCounters`:

https://github.com/kubernetes/kubernetes/blob/9c2cbda9d8d0e2e6eb43b58b2369b7be90bfe564/pkg/apis/resource/validation/validation_resourceslice_test.go#L425-L430

Referencing one counter in it:

https://github.com/kubernetes/kubernetes/blob/9c2cbda9d8d0e2e6eb43b58b2369b7be90bfe564/pkg/apis/resource/validation/validation_resourceslice_test.go#L447-L457

The `SharedCounter: fmt.Sprintf("set-%d", i)` line isn't right.

Reply:
No, the line is right! It just looks extremely odd because of the `SharedCounter` field name. That should be `CounterSet: fmt.Sprintf("set-%d", i)`.

> So it is a bit unclear is the plural in the SharedCounters name references the multiple CounterSets or the multiple Counters.

It's the "multiple counters". That's also how the size limit is now defined: the total number of counters in that field, not the number of different sets.

Reply:
I suggested earlier to make it a flat list with `ShareCounters []Counter`, with the name inside the struct, but @johnbelamaric convinced me that the two-level nesting with sets as outer layer is needed for mixins.

Reply:
Yeah, I know. I was trying to do to many things at once and realized I had misunderstood what I was reading..

Reply:
> It's the "multiple counters". That's also how the size limit is now defined: the total number of counters in that field, not the number of different sets.

Yeah, I think that makes sense. I will update the API to use the plural for `ConsumesCounter` and `SharedCounter` to `CounterSet`. I think we will just continue using `consumes` as I think the take-away from the discussion on the issue is that the alternatives aren't clearly better.